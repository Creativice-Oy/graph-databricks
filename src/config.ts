import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  databricksSubdomain: {
    type: 'string',
  },
  databricksAccessToken: {
    type: 'string',
    mask: true,
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  databricksSubdomain: string;
  databricksAccessToken: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.databricksSubdomain || !config.databricksAccessToken) {
    throw new IntegrationValidationError(
      'Config requires all of {databricksSubdomain, databricksAccessToken}',
    );
  }

  const apiClient = createAPIClient(config);
  await apiClient.verifyAuthentication();
}
