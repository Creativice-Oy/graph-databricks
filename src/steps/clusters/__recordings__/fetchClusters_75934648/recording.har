{
  "log": {
    "_recordingName": "fetchClusters",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "bf48bfc2950b87846e3443d5d700288c",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 320,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/clusters/list"
        },
        "response": {
          "bodySize": 1977,
          "content": {
            "mimeType": "application/json",
            "size": 1977,
            "text": "{\"clusters\":[{\"cluster_id\":\"0201-183853-7c8r8974\",\"driver\":{\"private_ip\":\"127.0.0.1\"},\"spark_context_id\":8745868079977237257,\"cluster_name\":\"test-cluster\",\"spark_version\":\"9.1.x-scala2.12\",\"spark_conf\":{\"spark.master\":\"local[*, 4]\",\"spark.databricks.cluster.profile\":\"singleNode\"},\"node_type_id\":\"n1-standard-4\",\"driver_node_type_id\":\"n1-standard-4\",\"custom_tags\":{\"ResourceClass\":\"SingleNode\"},\"spark_env_vars\":{\"PYSPARK_PYTHON\":\"/databricks/python3/bin/python3\"},\"autotermination_minutes\":20,\"enable_elastic_disk\":false,\"disk_spec\":{},\"cluster_source\":\"UI\",\"enable_local_disk_encryption\":false,\"gcp_attributes\":{\"use_preemptible_executors\":false,\"availability\":\"ON_DEMAND_GCP\",\"zone_id\":\"HA\"},\"instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"driver_instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"state\":\"RUNNING\",\"start_time\":1643740733901,\"last_state_loss_time\":1643741512861,\"num_workers\":0,\"cluster_memory_mb\":15360,\"cluster_cores\":4.0,\"default_tags\":{\"Vendor\":\"Databricks\",\"Creator\":\"viragsf_at_gmail.com\",\"ClusterName\":\"test-cluster\",\"ClusterId\":\"0201-183853-7c8r8974\"},\"creator_user_name\":\"viragsf@gmail.com\",\"init_scripts_safe_mode\":false},{\"cluster_id\":\"0117-142331-yanks462\",\"driver\":{\"private_ip\":\"127.0.0.1\"},\"cluster_name\":\"test-gcp-cluster\",\"spark_version\":\"9.1.x-scala2.12\",\"node_type_id\":\"n1-standard-4\",\"driver_node_type_id\":\"n1-standard-4\",\"spark_env_vars\":{\"PYSPARK_PYTHON\":\"/databricks/python3/bin/python3\"},\"autotermination_minutes\":20,\"enable_elastic_disk\":false,\"disk_spec\":{},\"cluster_source\":\"UI\",\"enable_local_disk_encryption\":false,\"gcp_attributes\":{\"use_preemptible_executors\":false,\"availability\":\"ON_DEMAND_GCP\",\"zone_id\":\"HA\"},\"instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"driver_instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"state\":\"TERMINATED\",\"state_message\":\"Cluster terminated by CLOUD_PROVIDER_LAUNCH_FAILURE\",\"start_time\":1642429411156,\"terminated_time\":1642429429719,\"last_state_loss_time\":1642429429732,\"autoscale\":{\"min_workers\":1,\"max_workers\":2},\"default_tags\":{\"Vendor\":\"Databricks\",\"Creator\":\"austin.kelleher_at_jupiterone.dev\",\"ClusterName\":\"test-gcp-cluster\",\"ClusterId\":\"0117-142331-yanks462\"},\"creator_user_name\":\"austin.kelleher@jupiterone.dev\",\"termination_reason\":{\"code\":\"CLOUD_PROVIDER_LAUNCH_FAILURE\",\"parameters\":{\"gcp_error_message\":\"Compute Quota Exceeded for jupiterone-databricks-test in region us-west1: Quota: SSD_TOTAL_GB, used 0.0 and requested 1000.0 out of 500.0\"},\"type\":\"CLOUD_FAILURE\"},\"init_scripts_safe_mode\":false}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 19:02:56 GMT"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 330,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T19:02:55.914Z",
        "time": 772,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 772
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
