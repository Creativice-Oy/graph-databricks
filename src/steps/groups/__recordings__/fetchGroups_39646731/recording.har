{
  "log": {
    "_recordingName": "fetchGroups",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "d31c258b4eecb943136691f6a1410ecb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "dbc-c50dbe80-ed72.cloud.databricks.com"
            }
          ],
          "headersSize": 320,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://dbc-c50dbe80-ed72.cloud.databricks.com/api/2.0/groups/list"
        },
        "response": {
          "bodySize": 27,
          "content": {
            "mimeType": "application/json",
            "size": 27,
            "text": "{\"group_names\":[\"admins\"]}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "date",
              "value": "Tue, 01 Mar 2022 17:20:53 GMT"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-1"
            },
            {
              "name": "content-length",
              "value": "27"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-1.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "1393914232801712"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 534,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-03-01T17:20:51.877Z",
        "time": 1571,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1571
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
