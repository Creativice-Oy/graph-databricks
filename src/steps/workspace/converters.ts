import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';

export function createWorkspaceEntity(subdomain: string): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: subdomain,
      },
      assign: {
        _key: `databricks_workspace:${subdomain}`,
        _type: Entities.WORKSPACE._type,
        _class: Entities.WORKSPACE._class,
        name: subdomain,
      },
    },
  });
}
