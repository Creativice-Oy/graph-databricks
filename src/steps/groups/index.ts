import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../../client';
import { IntegrationConfig } from '../../config';
import { WORKSPACE_ENTITY_KEY } from '../workspace';
import { Entities, Relationships, Steps } from '../constants';
import { createGroupEntity, createUserEntity } from './converters';

export async function fetchGroups({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const workspaceEntity = (await jobState.getData(
    WORKSPACE_ENTITY_KEY,
  )) as Entity;

  await apiClient.iterateGroups(async (group) => {
    const groupEntity = await jobState.addEntity(createGroupEntity(group));
    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: workspaceEntity,
        to: groupEntity,
      }),
    );
  });
}

export async function fetchGroupUsers({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  await jobState.iterateEntities(
    {
      _type: Entities.GROUP._type,
    },
    async (groupEntity) => {
      await apiClient.iterateGroupUsers(
        groupEntity.displayName!,
        async (user) => {
          const userEntity = await jobState.addEntity(createUserEntity(user));

          await jobState.addRelationship(
            createDirectRelationship({
              _class: RelationshipClass.HAS,
              from: groupEntity,
              to: userEntity,
            }),
          );
        },
      );
    },
  );
}

export const groupSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.GROUPS,
    name: 'Fetch Groups',
    entities: [Entities.GROUP],
    relationships: [Relationships.WORKSPACE_HAS_GROUP],
    dependsOn: [Steps.WORKSPACE],
    executionHandler: fetchGroups,
  },
  {
    id: Steps.GROUP_USERS,
    name: 'Build Group User Relationships',
    entities: [Entities.USER],
    relationships: [Relationships.GROUP_HAS_USER],
    dependsOn: [Steps.GROUPS],
    executionHandler: fetchGroupUsers,
  },
];
