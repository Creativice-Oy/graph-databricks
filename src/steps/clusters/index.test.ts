import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { buildClusterUserRelationships, fetchClusters } from '.';
import { fetchGroups, fetchGroupUsers } from '../groups';
import { integrationConfig } from '../../../test/config';
import { setupDatabricksRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../config';
import { Relationships } from '../constants';
import { fetchWorkspaceDetails } from '../workspace';

describe('#fetchClusters', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupDatabricksRecording({
      directory: __dirname,
      name: 'fetchClusters',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchWorkspaceDetails(context);
    await fetchClusters(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const clusters = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('databricks_cluster'),
    );

    expect(clusters.length).toBeGreaterThan(0);
    expect(clusters).toMatchGraphObjectSchema({
      _class: ['Cluster'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'databricks_cluster' },
          name: { type: 'string' },
          creator: { type: 'string' },
          state: { type: 'string' },
          active: { type: 'boolean' },
          autoTerminationMinutes: { type: 'number' },
          enableElasticDisk: { type: 'boolean' },
          enableLocalDiskEncryption: { type: 'boolean' },
          initScriptsSafeMode: { type: 'boolean' },
          nodeTypeId: { type: 'string' },
          createdOn: { type: 'number' },
          terminatedOn: { type: 'number' },
        },
      },
    });

    const workspaceHasClusters = context.jobState.collectedRelationships.filter(
      (e) => e._type === Relationships.WORKSPACE_HAS_CLUSTER._type,
    );

    expect(workspaceHasClusters.length).toBeGreaterThan(0);
    expect(workspaceHasClusters).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'databricks_workspace_has_cluster',
          },
        },
      },
    });
  });
});

describe('#buildClusterUserRelationships', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupDatabricksRecording({
      directory: __dirname,
      name: 'buildClusterUserRelationships',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchWorkspaceDetails(context);
    await fetchGroups(context);
    await fetchGroupUsers(context);
    await fetchClusters(context);
    await buildClusterUserRelationships(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const clusters = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('databricks_cluster'),
    );
    const users = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('databricks_user'),
    );

    expect(clusters.length).toBeGreaterThan(0);
    expect(users.length).toBeGreaterThan(0);

    const userCreatedClusters = context.jobState.collectedRelationships.filter(
      (e) => e._type === Relationships.USER_CREATED_CLUSTER._type,
    );

    expect(userCreatedClusters.length).toBeGreaterThan(0);
    expect(userCreatedClusters).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'CREATED' },
          _type: {
            const: 'databricks_user_created_cluster',
          },
        },
      },
    });
  });
});
