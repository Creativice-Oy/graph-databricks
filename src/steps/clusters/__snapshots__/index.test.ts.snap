// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildClusterUserRelationships should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "databricks_workspace:3840207185082406.6",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "3840207185082406.6",
          },
        },
      ],
      "_type": "databricks_workspace",
      "createdOn": undefined,
      "displayName": "3840207185082406.6",
      "id": "3840207185082406.6",
      "name": "3840207185082406.6",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "databricks_group:admins",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "admins",
          },
        },
      ],
      "_type": "databricks_group",
      "createdOn": undefined,
      "displayName": "admins",
      "name": "admins",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "databricks_group:test-group-1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "test-group-1",
          },
        },
      ],
      "_type": "databricks_group",
      "createdOn": undefined,
      "displayName": "test-group-1",
      "name": "test-group-1",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "databricks_group:test-group-2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "test-group-2",
          },
        },
      ],
      "_type": "databricks_group",
      "createdOn": undefined,
      "displayName": "test-group-2",
      "name": "test-group-2",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "databricks_user:viragsf@gmail.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "username": "viragsf@gmail.com",
          },
        },
      ],
      "_type": "databricks_user",
      "createdOn": undefined,
      "displayName": "viragsf@gmail.com",
      "name": "viragsf@gmail.com",
      "username": "viragsf@gmail.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "databricks_user:austin.kelleher@jupiterone.dev",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "username": "austin.kelleher@jupiterone.dev",
          },
        },
      ],
      "_type": "databricks_user",
      "createdOn": undefined,
      "displayName": "austin.kelleher@jupiterone.dev",
      "name": "austin.kelleher@jupiterone.dev",
      "username": "austin.kelleher@jupiterone.dev",
    },
    Object {
      "_class": Array [
        "Cluster",
      ],
      "_key": "databricks_cluster:0201-183853-7c8r8974",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "autotermination_minutes": 20,
            "cluster_cores": 4,
            "cluster_id": "0201-183853-7c8r8974",
            "cluster_memory_mb": 15360,
            "cluster_name": "test-cluster",
            "cluster_source": "UI",
            "creator_user_name": "viragsf@gmail.com",
            "custom_tags": Object {
              "ResourceClass": "SingleNode",
            },
            "default_tags": Object {
              "ClusterId": "0201-183853-7c8r8974",
              "ClusterName": "test-cluster",
              "Creator": "viragsf_at_gmail.com",
              "Vendor": "Databricks",
            },
            "disk_spec": Object {},
            "driver": Object {
              "private_ip": "127.0.0.1",
            },
            "driver_instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "driver_node_type_id": "n1-standard-4",
            "enable_elastic_disk": false,
            "enable_local_disk_encryption": false,
            "gcp_attributes": Object {
              "availability": "ON_DEMAND_GCP",
              "use_preemptible_executors": false,
              "zone_id": "HA",
            },
            "init_scripts_safe_mode": false,
            "instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "last_state_loss_time": 1643741512861,
            "node_type_id": "n1-standard-4",
            "num_workers": 0,
            "spark_conf": Object {
              "spark.databricks.cluster.profile": "singleNode",
              "spark.master": "local[*, 4]",
            },
            "spark_context_id": 8745868079977238000,
            "spark_env_vars": Object {
              "PYSPARK_PYTHON": "/databricks/python3/bin/python3",
            },
            "spark_version": "9.1.x-scala2.12",
            "start_time": 1643740733901,
            "state": "RUNNING",
          },
        },
      ],
      "_type": "databricks_cluster",
      "active": true,
      "autoTerminationMinutes": 20,
      "createdOn": 1643740733901,
      "creator": "viragsf@gmail.com",
      "displayName": "test-cluster",
      "enableElasticDisk": false,
      "enableLocalDiskEncryption": false,
      "initScriptsSafeMode": false,
      "name": "test-cluster",
      "nodeTypeId": "n1-standard-4",
      "state": "RUNNING",
      "terminatedOn": undefined,
    },
    Object {
      "_class": Array [
        "Cluster",
      ],
      "_key": "databricks_cluster:0117-142331-yanks462",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "autoscale": Object {
              "max_workers": 2,
              "min_workers": 1,
            },
            "autotermination_minutes": 20,
            "cluster_id": "0117-142331-yanks462",
            "cluster_name": "test-gcp-cluster",
            "cluster_source": "UI",
            "creator_user_name": "austin.kelleher@jupiterone.dev",
            "default_tags": Object {
              "ClusterId": "0117-142331-yanks462",
              "ClusterName": "test-gcp-cluster",
              "Creator": "austin.kelleher_at_jupiterone.dev",
              "Vendor": "Databricks",
            },
            "disk_spec": Object {},
            "driver": Object {
              "private_ip": "127.0.0.1",
            },
            "driver_instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "driver_node_type_id": "n1-standard-4",
            "enable_elastic_disk": false,
            "enable_local_disk_encryption": false,
            "gcp_attributes": Object {
              "availability": "ON_DEMAND_GCP",
              "use_preemptible_executors": false,
              "zone_id": "HA",
            },
            "init_scripts_safe_mode": false,
            "instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "last_state_loss_time": 1642429429732,
            "node_type_id": "n1-standard-4",
            "spark_env_vars": Object {
              "PYSPARK_PYTHON": "/databricks/python3/bin/python3",
            },
            "spark_version": "9.1.x-scala2.12",
            "start_time": 1642429411156,
            "state": "TERMINATED",
            "state_message": "Cluster terminated by CLOUD_PROVIDER_LAUNCH_FAILURE",
            "terminated_time": 1642429429719,
            "termination_reason": Object {
              "code": "CLOUD_PROVIDER_LAUNCH_FAILURE",
              "parameters": Object {
                "gcp_error_message": "Compute Quota Exceeded for jupiterone-databricks-test in region us-west1: Quota: SSD_TOTAL_GB, used 0.0 and requested 1000.0 out of 500.0",
              },
              "type": "CLOUD_FAILURE",
            },
          },
        },
      ],
      "_type": "databricks_cluster",
      "active": false,
      "autoTerminationMinutes": 20,
      "createdOn": 1642429411156,
      "creator": "austin.kelleher@jupiterone.dev",
      "displayName": "test-gcp-cluster",
      "enableElasticDisk": false,
      "enableLocalDiskEncryption": false,
      "initScriptsSafeMode": false,
      "name": "test-gcp-cluster",
      "nodeTypeId": "n1-standard-4",
      "state": "TERMINATED",
      "terminatedOn": 1642429429719,
    },
  ],
  "encounteredTypes": Array [
    "databricks_workspace",
    "databricks_group",
    "databricks_workspace_has_group",
    "databricks_user",
    "databricks_group_has_user",
    "databricks_cluster",
    "databricks_workspace_has_cluster",
    "databricks_user_created_cluster",
  ],
  "numCollectedEntities": 8,
}
`;

exports[`#fetchClusters should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "databricks_workspace:3840207185082406.6",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "3840207185082406.6",
          },
        },
      ],
      "_type": "databricks_workspace",
      "createdOn": undefined,
      "displayName": "3840207185082406.6",
      "id": "3840207185082406.6",
      "name": "3840207185082406.6",
    },
    Object {
      "_class": Array [
        "Cluster",
      ],
      "_key": "databricks_cluster:0201-183853-7c8r8974",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "autotermination_minutes": 20,
            "cluster_cores": 4,
            "cluster_id": "0201-183853-7c8r8974",
            "cluster_memory_mb": 15360,
            "cluster_name": "test-cluster",
            "cluster_source": "UI",
            "creator_user_name": "viragsf@gmail.com",
            "custom_tags": Object {
              "ResourceClass": "SingleNode",
            },
            "default_tags": Object {
              "ClusterId": "0201-183853-7c8r8974",
              "ClusterName": "test-cluster",
              "Creator": "viragsf_at_gmail.com",
              "Vendor": "Databricks",
            },
            "disk_spec": Object {},
            "driver": Object {
              "private_ip": "127.0.0.1",
            },
            "driver_instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "driver_node_type_id": "n1-standard-4",
            "enable_elastic_disk": false,
            "enable_local_disk_encryption": false,
            "gcp_attributes": Object {
              "availability": "ON_DEMAND_GCP",
              "use_preemptible_executors": false,
              "zone_id": "HA",
            },
            "init_scripts_safe_mode": false,
            "instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "last_state_loss_time": 1643741512861,
            "node_type_id": "n1-standard-4",
            "num_workers": 0,
            "spark_conf": Object {
              "spark.databricks.cluster.profile": "singleNode",
              "spark.master": "local[*, 4]",
            },
            "spark_context_id": 8745868079977238000,
            "spark_env_vars": Object {
              "PYSPARK_PYTHON": "/databricks/python3/bin/python3",
            },
            "spark_version": "9.1.x-scala2.12",
            "start_time": 1643740733901,
            "state": "RUNNING",
          },
        },
      ],
      "_type": "databricks_cluster",
      "active": true,
      "autoTerminationMinutes": 20,
      "createdOn": 1643740733901,
      "creator": "viragsf@gmail.com",
      "displayName": "test-cluster",
      "enableElasticDisk": false,
      "enableLocalDiskEncryption": false,
      "initScriptsSafeMode": false,
      "name": "test-cluster",
      "nodeTypeId": "n1-standard-4",
      "state": "RUNNING",
      "terminatedOn": undefined,
    },
    Object {
      "_class": Array [
        "Cluster",
      ],
      "_key": "databricks_cluster:0117-142331-yanks462",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "autoscale": Object {
              "max_workers": 2,
              "min_workers": 1,
            },
            "autotermination_minutes": 20,
            "cluster_id": "0117-142331-yanks462",
            "cluster_name": "test-gcp-cluster",
            "cluster_source": "UI",
            "creator_user_name": "austin.kelleher@jupiterone.dev",
            "default_tags": Object {
              "ClusterId": "0117-142331-yanks462",
              "ClusterName": "test-gcp-cluster",
              "Creator": "austin.kelleher_at_jupiterone.dev",
              "Vendor": "Databricks",
            },
            "disk_spec": Object {},
            "driver": Object {
              "private_ip": "127.0.0.1",
            },
            "driver_instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "driver_node_type_id": "n1-standard-4",
            "enable_elastic_disk": false,
            "enable_local_disk_encryption": false,
            "gcp_attributes": Object {
              "availability": "ON_DEMAND_GCP",
              "use_preemptible_executors": false,
              "zone_id": "HA",
            },
            "init_scripts_safe_mode": false,
            "instance_source": Object {
              "node_type_id": "n1-standard-4",
            },
            "last_state_loss_time": 1642429429732,
            "node_type_id": "n1-standard-4",
            "spark_env_vars": Object {
              "PYSPARK_PYTHON": "/databricks/python3/bin/python3",
            },
            "spark_version": "9.1.x-scala2.12",
            "start_time": 1642429411156,
            "state": "TERMINATED",
            "state_message": "Cluster terminated by CLOUD_PROVIDER_LAUNCH_FAILURE",
            "terminated_time": 1642429429719,
            "termination_reason": Object {
              "code": "CLOUD_PROVIDER_LAUNCH_FAILURE",
              "parameters": Object {
                "gcp_error_message": "Compute Quota Exceeded for jupiterone-databricks-test in region us-west1: Quota: SSD_TOTAL_GB, used 0.0 and requested 1000.0 out of 500.0",
              },
              "type": "CLOUD_FAILURE",
            },
          },
        },
      ],
      "_type": "databricks_cluster",
      "active": false,
      "autoTerminationMinutes": 20,
      "createdOn": 1642429411156,
      "creator": "austin.kelleher@jupiterone.dev",
      "displayName": "test-gcp-cluster",
      "enableElasticDisk": false,
      "enableLocalDiskEncryption": false,
      "initScriptsSafeMode": false,
      "name": "test-gcp-cluster",
      "nodeTypeId": "n1-standard-4",
      "state": "TERMINATED",
      "terminatedOn": 1642429429719,
    },
  ],
  "encounteredTypes": Array [
    "databricks_workspace",
    "databricks_cluster",
    "databricks_workspace_has_cluster",
  ],
  "numCollectedEntities": 3,
}
`;
